[Clang] [C++26] Expansion Statements (Part 6: Destructuring Expansion Statements)

Sirraide opened the pull request:
This implements Sema and template instantiation for destructuring expansion statements.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 6)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 6)

erichkeane reviewed (approved):
would be nice to be better about how we pass around/come up with our ExpressionEvaluationContext variables in these patches, else LGTM.

Sirraide commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -274,6 +274,51 @@ static IterableExpansionStmtData TryBuildIterableExpansionStmtInitializer(
+    ArrayRef<MaterializeTemporaryExpr *> LifetimeExtendTemps) {
+  auto Ctx = Sema::ExpressionEvaluationContext::PotentiallyEvaluated;
+  if (VarIsConstexpr)
+    Ctx = Sema::ExpressionEvaluationContext::ImmediateFunctionContext;
```
So at least when it comes to this file, `VarIsConstexpr` is used for more than just deciding what expression evaluation context to push (e.g. it’s used to make the `__range` variable `constexpr` iff it is declared `constexpr`), and I feel like passing the expression evaluation context enum as a parameter would make it a lot less obvious as to what’s going on there

cor3ntin replied:
We could have a `Sema::ExpressionEvaluationContextEvaluatedOrImmediate(bool constexpr)`.

However in this context shouldn't the context always be immediate?

Sirraide replied:
I don’t think so; since the standard specifies that we should generate
```c++
auto [_U1, _U2/*, ...*/] = expansion-initializer;
```
I don’t think this should be a constant evaluated / immediate context if the variable is not declared `constexpr`

cor3ntin replied:
hum, nvm that bit, as long as the size is a constexpr expression it's fine - my suggestion for a function stands

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3738,6 +3738,8 @@ def err_conflicting_codeseg_attribute : Error<
 def warn_duplicate_codeseg_attribute : Warning<
   "duplicate code segment specifiers">, InGroup<Section>;
 
+def err_expansion_stmt_invalid_init : Error<
+  "cannot expand expression of type %0">;
```
Can we massage that a bit? Maybe "expression of type %0 cannot be destructured".

Sirraide replied:
Hmm, I don’t think that wording would be a good idea: the issue here isn’t really that it can’t be destructured; it’s that it cannot be destructured _and_ also isn’t iterable. An expression that can be expanded need not be destructurable and vice versa

cor3ntin replied:
Hum - right - but in that case can you move the emission of the diagnostic outside of `BuildDestructuringDecompositionDecl` -  ideally in the Act function? So that the flow is logical.

Thanks

Sirraide replied:
> ideally in the Act function?

That’s not possible, but I’ll move it into `BuildNonEnumeratingCXXExpansionStmtPattern()`; that’s the best I can do I think

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -224,6 +224,51 @@ static IterableExpansionStmtData TryBuildIterableExpansionStmtInitializer(
+  for (unsigned I = 0; I < *Arity; ++I)
+    Bindings.push_back(BindingDecl::Create(
+        S.Context, S.CurContext, ColonLoc,
+        S.getPreprocessor().getIdentifierInfo("__u" + std::to_string(I)),
```
You should not need a name here

Sirraide replied:
Hmm, I wonder how that would affect diagnostics involving this binding; if we ever try to print the variable name I’m not sure what the user would see

cor3ntin replied:
Are we ever doing that?

Sirraide replied:
Yeah, without the name, we get ‘constexpr variable '[]' must be initialized by a constant expression’... I guess we can try and update all diagnostics that may reference a variable name to drop it if it’s empty, but I feel like it’d be easier to just make up a name here

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -224,6 +224,51 @@ static IterableExpansionStmtData TryBuildIterableExpansionStmtInitializer(
+      DecompositionDecl::Create(S.Context, S.CurContext, ColonLoc, ColonLoc,
+                                AutoRRef, TSI, SC_Auto, Bindings);
+
+  if (VarIsConstexpr)
+    DD->setConstexpr(true);
```
Again, don't we *need* the variable to be constexpr?

Sirraide replied:
This was changed over the course of a few core issues; as of now, the variable is constexpr iff the user actually wrote `constexpr` (if you look at [[stmt.expand]](https://eel.is/c++draft/stmt.expand), you’ll see that it has ‘constexpr<sub>_opt_</sub>’ everywhere)

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -384,8 +429,59 @@ StmtResult Sema::BuildNonEnumeratingCXXExpansionStmtPattern(
+  // Synthesise an InitListExpr to store the bindings; this essentially lets us
+  // desugar the expansion of a destructuring expansion statement to that of an
+  // enumerating expansion statement.
+  SmallVector<Expr *> Bindings;
```
Can you call reserve?

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -384,8 +429,59 @@ StmtResult Sema::BuildNonEnumeratingCXXExpansionStmtPattern(
+  // enumerating expansion statement.
+  SmallVector<Expr *> Bindings;
+  for (BindingDecl *BD : DD->bindings()) {
+    Expr *Element = BuildDeclRefExpr(BD, BD->getType().getNonReferenceType(),
+                                     VK_LValue, ColonLoc);
+
+    // CWG 3149: If the expansion-initializer is an lvalue, then vi is ui;
+    // otherwise, vi is static_cast<decltype(ui)&&>(ui).
+    if (!ExpansionInitializer->isLValue()) {
+      QualType Ty =
+          BuildReferenceType(getDecltypeForExpr(Element), /*LValueRef=*/false,
+                             ColonLoc, /*Entity=*/DeclarationName());
+      TypeSourceInfo *TSI = Context.getTrivialTypeSourceInfo(Ty);
+      ExprResult Cast = BuildCXXNamedCast(
+          ColonLoc, tok::kw_static_cast, TSI, Element,
+          SourceRange(ColonLoc, ColonLoc), SourceRange(ColonLoc, ColonLoc));
+      assert(!Cast.isInvalid() && "cast to rvalue reference type failed?");
+      Element = Cast.get();
```
Do we actually have to do all that, or can we create `BuildDeclRefExpr` with an rvalue?

Sirraide replied:
I never realised you could do that; I’ll try and see if that works

Sirraide replied:
This doesn’t work; a DRE that is not an `LValue` or `PRValue` causes some assertions to fail

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -9479,7 +9479,12 @@ StmtResult TreeTransform<Derived>::TransformCXXExpansionStmtPattern(
 
     NewPattern = cast<CXXExpansionStmtPattern>(Res.get());
   } else {
-    llvm_unreachable("TODO");
+    // The only time we instantiate an expansion statement is if its expansion
+    // size is dependent (otherwise, we only instantiate the expansions and
+    // leave the underlying CXXExpansionStmtPattern as-is). Since destructuring
+    // expansion statements never have a dependent size, we should never get
+    // here.
+    llvm_unreachable("destructuring pattern should never be instantiated");
```
tuple_size can certainly be dependent - Am i missunderstanding?

Sirraide replied:
I don’t believe that is possible unless the _expansion-initializer_ is dependent—but if it is... we don’t build a ‘destructuring’ `CXXExpansionStmt` but rather a ‘dependent’ one. Is there any case I’m missing here?

shafik commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3738,6 +3738,8 @@ def err_conflicting_codeseg_attribute : Error<
   "duplicate code segment specifiers">, InGroup<Section>;
 
+def err_expansion_stmt_invalid_init : Error<
+  "cannot expand expression of type %0">;
```
Which PR has the test that covers this diagnostic?

Sirraide replied:
7

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -234,6 +234,46 @@ static IterableExpansionStmtData TryBuildIterableExpansionStmtInitializer(
+  for (unsigned I = 0; I < *Arity; ++I)
+    Bindings.push_back(BindingDecl::Create(
+        S.Context, S.CurContext, ColonLoc,
+        S.getPreprocessor().getIdentifierInfo("__u" + std::to_string(I)),
```
So naming these things has me asking myself some questions. I think naming them does make sense but what happens if we have a local inside a expansion and it has the same name will they conflict? 

How does the line numbering look for a diagnostic?

I would like to see tests of this scenario with a local inside the expansion with a name matching this. Yes, it would be IFNDR but we should test it.

shafik replied:
I would also be curious how this works w/ nested expansion statements, that I am less worried about but I would be curious to see if the diagnostics are intelligible.

Sirraide replied:
The body is in a separate scope, and I don’t think these declarations are even added to any scopes to begin with, so nested expansion statements shouldn’t be an issue either, but I’ll add a test with `-Wshadow` enabled and see if that fires (I doubt it will but still)

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -388,8 +428,65 @@ StmtResult Sema::BuildNonEnumeratingCXXExpansionStmtPattern(
+  // desugar the expansion of a destructuring expansion statement to that of an
+  // enumerating expansion statement.
+  SmallVector<Expr *> Bindings;
+  Bindings.reserve(DD->bindings().size());
```
I almost wonder if this is a premature optimization, I don't think we have data one way or the other.

Sirraide replied:
Corentin asked me to add the call to `reserve()`

shafik replied:
I saw that, it is not a big deal but it smells like premature opt @cor3ntin but this is not the biggest fish to fry.

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -388,8 +428,65 @@ StmtResult Sema::BuildNonEnumeratingCXXExpansionStmtPattern(
+    Expr *Element = BuildDeclRefExpr(BD, BD->getType().getNonReferenceType(),
+                                     VK_LValue, ColonLoc);
+
+    // CWG 3149: If the expansion-initializer is an lvalue, then vi is ui;
```
I think we should also quote `[stmt.expand]` since this was accepted. I don't mind also quoting the core issue but if we have the wording in the standard we should reference that with a reference and a quote.

This applies to anywhere else you quoted an accepted core issue.

Sirraide replied:
Yes, these comments predate the core issues being accepted

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 7: Constexpr support and tests) (#169686)

This implements compile-time evaluation of expansion statements. With
this implemented, we can finally test expansion statements properly, so
this also adds a bunch of tests for all three kinds of expansion
statements.

I didn’t add any Sema tests before this because I felt that just testing
that we don’t emit any diagnostics would not have been too useful; with
this patch, we can also test that the behaviour of the generated AST is
correct by evaluating it.
